{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bellfort Sequence Parser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tkinter as tk\n",
    "from tkinter import ttk\n",
    "import tkinter.font as tkf\n",
    "from tkinter import messagebox\n",
    "from tkinter import filedialog\n",
    "import threading\n",
    "import time\n",
    "import os\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions\n",
    "### Reverse Complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def reverseComplement(sequence):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
    "    rc_sequence=''\n",
    "    for s in sequence:\n",
    "        rc_sequence = complement[s] + rc_sequence\n",
    "    return rc_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTQ File Browse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonBrowseFASTQ():\n",
    "    global filenameFASTQ\n",
    "    \n",
    "    try:\n",
    "        filenameFASTQ = filedialog.askopenfilename(filetypes=(('FASTQ files', '*.fastq'), \n",
    "                                                              ('All files', '*.*')))\n",
    "        text_fileFASTQ.delete('1.0', tk.END)\n",
    "        text_fileFASTQ.insert('1.0', filenameFASTQ.split('/')[-1])\n",
    "    except:\n",
    "        filenameFASTQ = ''   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTQ File Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def loadFASTQ():\n",
    "    global reads\n",
    "    \n",
    "    start_time = time.time()    \n",
    "    \n",
    "    f = open(filenameFASTQ)\n",
    "\n",
    "    reads = []\n",
    "\n",
    "    try:\n",
    "        while 1:\n",
    "            name = f.readline().rstrip()\n",
    "            sequence = f.readline().rstrip()\n",
    "            f.readline()\n",
    "            quality = f.readline().rstrip()\n",
    "\n",
    "            if len(name) == 0:\n",
    "                break\n",
    "\n",
    "            union = name, sequence\n",
    "\n",
    "            reads.append(union)           \n",
    "\n",
    "        end_time = time.time()\n",
    "        delta_time = end_time - start_time\n",
    "\n",
    "        text_time.delete('1.0', tk.END)\n",
    "        text_time.insert('1.0', str(delta_time))  \n",
    "\n",
    "        text_readNum.delete('1.0', tk.END)\n",
    "        text_readNum.insert('1.0', str(len(reads)))  \n",
    "\n",
    "    except:\n",
    "        messagebox.showwarning(\"File Loading Failed\", \n",
    "                               \"Sorry, file loading failed! Please check the file format.\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def start_loadFASTQ_thread(event):\n",
    "    global loadFASTQ_thread\n",
    "    \n",
    "    if filenameFASTQ != '':\n",
    "        loadFASTQ_thread = threading.Thread(target=loadFASTQ)\n",
    "        loadFASTQ_thread.daemon = True\n",
    "\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        loadFASTQ_thread.start()\n",
    "        root.after(20, check_loadFASTQ_thread)\n",
    "    else:\n",
    "        messagebox.showwarning(\"No File\", \n",
    "                               \"Sorry, no file loaded! Please choose FASTQ file first.\")\n",
    "\n",
    "def check_loadFASTQ_thread():\n",
    "    if loadFASTQ_thread.is_alive():\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        root.after(20, check_loadFASTQ_thread)\n",
    "    else:\n",
    "        progressbar_loadFASTQ.stop()\n",
    "        progressbar_loadFASTQ['value']=100\n",
    "        messagebox.showinfo(\"FASTQ File Loaded\", \"FASTQ file successfully loaded!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Divide FASTQ File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def divideFASTQ():    \n",
    "    start_time = time.time() \n",
    "\n",
    "    gotten = text_readNumDivided.get('1.0', tk.END)\n",
    "    readNumDivided = int(gotten.rstrip())\n",
    "\n",
    "    if os.path.exists(filenameFASTQ+'.folder'):           \n",
    "        # Remove the folder previously made:\n",
    "        shutil.rmtree(filenameFASTQ+'.folder')\n",
    "\n",
    "    # Make a new one:\n",
    "    os.makedirs(filenameFASTQ+'.folder')    \n",
    "\n",
    "    line_num = 0\n",
    "    file_no = 1\n",
    "\n",
    "    f_input = open(filenameFASTQ)        \n",
    "    f_output = open(filenameFASTQ+'.folder/' + 'Reads_Slice_No_' + str(file_no) + '.fastq', 'w') \n",
    "\n",
    "    while 1:\n",
    "        # Input ///////////////////////////////////\n",
    "        name = f_input.readline()\n",
    "        sequence = f_input.readline()\n",
    "        f_input.readline()\n",
    "        quality = f_input.readline()\n",
    "\n",
    "        if len(name) == 0:\n",
    "            break            \n",
    "\n",
    "        # Output ////////////////////////////////////\n",
    "\n",
    "        f_output.write(name)\n",
    "        f_output.write(sequence)\n",
    "        f_output.write('+\\n')\n",
    "        f_output.write(quality)      \n",
    "\n",
    "        line_num += 1\n",
    "\n",
    "        if line_num == readNumDivided:                \n",
    "            f_output.close() \n",
    "            file_no += 1\n",
    "            f_output = open(filenameFASTQ+'.folder/' + 'Reads_Slice_No_' + str(file_no) + '.fastq', 'w')\n",
    "            line_num = 0                \n",
    "\n",
    "    end_time = time.time()\n",
    "    delta_time = end_time - start_time\n",
    "\n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))  \n",
    "\n",
    "    f_input.close()\n",
    "    f_output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_divideFASTQ_thread(event):\n",
    "    global divideFASTQ_thread\n",
    "    \n",
    "    if filenameFASTQ != '':\n",
    "        divideFASTQ_thread = threading.Thread(target=divideFASTQ)\n",
    "        divideFASTQ_thread.daemon = True\n",
    "\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        divideFASTQ_thread.start()\n",
    "        root.after(20, check_divideFASTQ_thread)\n",
    "    else:\n",
    "        messagebox.showwarning(\"No File\", \n",
    "                               \"Sorry, no file loaded! Please choose FASTQ file first.\")\n",
    "\n",
    "def check_divideFASTQ_thread():\n",
    "    if divideFASTQ_thread.is_alive():\n",
    "        progressbar_loadFASTQ.start(10)\n",
    "        root.after(20, check_divideFASTQ_thread)\n",
    "    else:\n",
    "        progressbar_loadFASTQ.stop()\n",
    "        progressbar_loadFASTQ['value']=100\n",
    "        messagebox.showinfo(\"FASTQ File Divided\", \"FASTQ file has been successfully divided!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def preprocessFASTQ():\n",
    "    global reads, indicator_preprocess, kmer_dict_reads\n",
    "    \n",
    "    try:\n",
    "        num = len(reads)   \n",
    "        indicator_preprocess = 0\n",
    "        gain = 500000/num\n",
    "\n",
    "        gotten = text_sequence_len.get('1.0', tk.END)\n",
    "        k = int(gotten.rstrip())\n",
    "        \n",
    "        if k > len(reads[0][1]):\n",
    "            messagebox.showwarning(\"Target Sequence Length Error\", \n",
    "                                   \"Sorry, the target sequence length is more than read length. Please check.\")\n",
    "        elif k < 3:\n",
    "            messagebox.showwarning(\"Sequence Too Short\", \n",
    "                                   \"Sorry, the target sequence length is too short which will make the program running slowly. Please check.\")\n",
    "        else:\n",
    "            kmer_dict_reads = {}\n",
    "\n",
    "            start_time = time.time()\n",
    "\n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]] = set()\n",
    "                indicator_preprocess += gain \n",
    "\n",
    "            for read in reads:\n",
    "                for i in range(len(read[1])-k+1):\n",
    "                    kmer_dict_reads[read[1][i:i+k]].add(read)\n",
    "                indicator_preprocess += gain\n",
    "\n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "\n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "\n",
    "            messagebox.showinfo(\"Preprocess FASTQ Completed\", \n",
    "                                \"Current FASTQ preprocess successfully completed!\")\n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ File Loaded\", \n",
    "                               \"Sorry, no loaded FASTQ file found! Please load FASTQ file first.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_preprocess_thread(event):\n",
    "    global preprocess_thread, indicator_preprocess\n",
    "    preprocess_thread = threading.Thread(target=preprocessFASTQ)\n",
    "    preprocess_thread.daemon = True\n",
    "    \n",
    "    progressbar['value'] = indicator_preprocess\n",
    "    \n",
    "    preprocess_thread.start()\n",
    "    root.after(20, check_preprocess_thread)\n",
    "\n",
    "def check_preprocess_thread():\n",
    "    if preprocess_thread.is_alive():\n",
    "        progressbar['value'] = indicator_preprocess\n",
    "        \n",
    "        root.after(20, check_preprocess_thread)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def matchAll():\n",
    "    global  kmer_dict_reads, indicator_matchAll, df\n",
    "    \n",
    "    try:\n",
    "        len(kmer_dict_reads)    \n",
    "        num = len(df)\n",
    "        \n",
    "        if num == 0:\n",
    "            messagebox.showwarning(\"No Sequences Loaded\", \n",
    "                                   \"Sorry, no sequences loaded! Please load sequences first.\")\n",
    "        else:    \n",
    "            indicator_matchAll = 0\n",
    "            gain = 1000000/num\n",
    "\n",
    "            start_time = time.time()\n",
    "\n",
    "            arr = np.array(df)\n",
    "\n",
    "            for i in range(len(arr)):\n",
    "                key1 = arr[i,2]\n",
    "                key2 = reverseComplement(key1)\n",
    "                \n",
    "                try:\n",
    "                    n1 = len(kmer_dict_reads[key1])\n",
    "                except KeyError:\n",
    "                    n1 = 0\n",
    "                    \n",
    "                try:\n",
    "                    n2 = len(kmer_dict_reads[key2])\n",
    "                except KeyError:\n",
    "                    n2 = 0\n",
    "                    \n",
    "                arr[i, 4] += n1 + n2\n",
    "                arr[i, 5] += 1\n",
    "                \n",
    "                indicator_matchAll += gain\n",
    "\n",
    "            df = pd.DataFrame(arr, columns = ['gene_id', 'UID', 'seq', 'Reserved', 'Count', 'Tag'])\n",
    "            #df = df.set_index('UID', drop=False) \n",
    "\n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "\n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "\n",
    "            messagebox.showinfo(\"Matching Completed\", \n",
    "                                \"Counting of sequences matched successfully completed!\")\n",
    "\n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No FASTQ Preprocessed or No Sequences Loaded\", \n",
    "                               \"Sorry, no FASTQ preprocess implemented or no sequences file loaded! Please preprocess FASTQ or load sequences first.\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def start_matchAll_thread(event):\n",
    "    global matchAll_thread, indicator_matchAll\n",
    "    matchAll_thread = threading.Thread(target=matchAll)\n",
    "    matchAll_thread.daemon = True\n",
    "    \n",
    "    progressbar['value'] = indicator_matchAll\n",
    "    \n",
    "    matchAll_thread.start()\n",
    "    root.after(20, check_matchAll_thread)\n",
    "\n",
    "def check_matchAll_thread():\n",
    "    if matchAll_thread.is_alive():\n",
    "        progressbar['value'] = indicator_matchAll\n",
    "        \n",
    "        root.after(20, check_matchAll_thread)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match Single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonMatch():\n",
    "    gotten = text_sequence.get('1.0', tk.END)\n",
    "    p1 = gotten.rstrip()    \n",
    "    p2 = reverseComplement(p1)\n",
    "    \n",
    "    if p1 == '' or p2 == '':\n",
    "        messagebox.showwarning(\"No Sequence Found\", \n",
    "                               \"Sorry, no sequence found in the text blank above! Please check the sequence.\")\n",
    "    else:\n",
    "        try:\n",
    "            len(kmer_dict_reads)\n",
    "            try:\n",
    "                n1 = len(kmer_dict_reads[p1])\n",
    "            except KeyError:\n",
    "                n1 = 0\n",
    "            \n",
    "            try:\n",
    "                n2 = len(kmer_dict_reads[p2])\n",
    "            except KeyError:\n",
    "                n2 = 0\n",
    "                \n",
    "            count = n1 + n2\n",
    "                \n",
    "            text_count.delete('1.0', tk.END)\n",
    "            text_count.insert('1.0', str(count))\n",
    "            \n",
    "        except NameError:\n",
    "            messagebox.showwarning(\"No FASTQ Preprocessed\", \n",
    "                                   \"Sorry, no FASTQ preprocess implemented! Please preprocess FASTQ first.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File of Target Sequence Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonBrowseSequences():\n",
    "    global filenameSequences\n",
    "    progressbar_loadSequences['value'] = 0\n",
    "    try:\n",
    "        filenameSequences = filedialog.askopenfilename(filetypes=(('Comma-Separated (CSV) text file', '*.csv'), ('All files', '*.*')))\n",
    "        text_fileSequences.delete('1.0', tk.END)\n",
    "        text_fileSequences.insert('1.0', filenameSequences.split('/')[-1])\n",
    "    except:\n",
    "        filenameSequences = ''    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def loadSequences():\n",
    "    global filenameSequences, df, recordNum\n",
    "   \n",
    "    if filenameSequences == '':\n",
    "        messagebox.showwarning(\"No File\", \"Sorry, no file chosen! Please choose file of sequences first.\")\n",
    "    else:        \n",
    "        try:\n",
    "            start_time = time.time()\n",
    "            \n",
    "            df = pd.read_csv(filenameSequences)\n",
    "            df['count'] = 0\n",
    "            df['tag'] = 0\n",
    "            #df = df.set_index('UID', drop=False)  \n",
    "            \n",
    "            recordNum = len(df)\n",
    "            \n",
    "            progressbar_loadSequences['value'] = 100\n",
    "            \n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "                       \n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "            \n",
    "            text_recordNum.delete('1.0', tk.END)\n",
    "            text_recordNum.insert('1.0', str(recordNum))\n",
    "            \n",
    "            messagebox.showinfo(\"File of Sequences Loaded\", \"File of sequences successfully loaded!\")        \n",
    "        except:\n",
    "            messagebox.showwarning(\"File Loading Failed\", \"Sorry, file loading failed! Please check the file format.\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonLoadHalfMatchedSequences():\n",
    "    global df, recordNum\n",
    "    \n",
    "    progressbar_loadSequences['value'] = 0\n",
    "    try:\n",
    "        filenameSequences = filedialog.askopenfilename(filetypes=(('Comma-Separated (CSV) text file', '*.csv'), ('All files', '*.*')))\n",
    "        text_fileSequences.delete('1.0', tk.END)\n",
    "        text_fileSequences.insert('1.0', filenameSequences.split('/')[-1])\n",
    "    except:\n",
    "        filenameSequences = ''    \n",
    "        \n",
    "    if filenameSequences == '':\n",
    "        messagebox.showwarning(\"No File\", \"Sorry, no file chosen! Please choose file of sequences first.\")\n",
    "    else:        \n",
    "        try:\n",
    "            start_time = time.time()\n",
    "            \n",
    "            df = pd.read_csv(filenameSequences)   \n",
    "            df = df.set_index('Unnamed: 0', drop=True)  \n",
    "            \n",
    "            recordNum = len(df)\n",
    "            \n",
    "            progressbar_loadSequences['value'] = 100\n",
    "            \n",
    "            end_time = time.time()\n",
    "            delta_time = end_time - start_time\n",
    "                       \n",
    "            text_time.delete('1.0', tk.END)\n",
    "            text_time.insert('1.0', str(delta_time))\n",
    "            \n",
    "            text_recordNum.delete('1.0', tk.END)\n",
    "            text_recordNum.insert('1.0', str(recordNum))\n",
    "            \n",
    "            messagebox.showinfo(\"File of Half Matched Sequences Loaded\", \"File of half matched sequences successfully loaded!\")        \n",
    "        except:\n",
    "            messagebox.showwarning(\"File Loading Failed\", \"Sorry, file loading failed! Please check the file format.\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>UID</th>\n",
       "      <th>seq</th>\n",
       "      <th>Reserved</th>\n",
       "      <th>Count</th>\n",
       "      <th>Tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGLibB_00001</td>\n",
       "      <td>CAATGTGACGCTGACGTGCC</td>\n",
       "      <td>HGLibB_00076</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGLibB_00002</td>\n",
       "      <td>TGTCTCCGCAGGTGTCACCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGLibB_00003</td>\n",
       "      <td>CATATGGCTGCTTCTGTCAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>HGLibB_00004</td>\n",
       "      <td>GGTGCAGCATCCCAACCAGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>HGLibB_00005</td>\n",
       "      <td>GGGATGCTGCACCCCCTGAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>HGLibB_00006</td>\n",
       "      <td>AACTTAATAATTATGAAATT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HGLibB_00007</td>\n",
       "      <td>CCCAAAGTCTTCATCCAATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HGLibB_00008</td>\n",
       "      <td>CCCACAGACGCCTCAGTCTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HGLibB_00009</td>\n",
       "      <td>TCCCAGGCAGTATATGGTTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>HGLibB_00010</td>\n",
       "      <td>GACAAACCTCTCTACACCCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>HGLibB_00011</td>\n",
       "      <td>GCCGGGCTGGTAATGTCACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>HGLibB_00012</td>\n",
       "      <td>GTATTTCCGCATTGTCACCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>HGLibB_00013</td>\n",
       "      <td>AAGTTCTGACCTGTACCCCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>HGLibB_00014</td>\n",
       "      <td>GGCCGAAGGGCAGACGCCCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>HGLibB_00015</td>\n",
       "      <td>AGCAAAGATAGTCAGCCCAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>HGLibB_00016</td>\n",
       "      <td>CGTGCGTCGTGGGGCAGTAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>HGLibB_00017</td>\n",
       "      <td>CTTCCACATGAGTGCGATCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>HGLibB_00018</td>\n",
       "      <td>GAACAATGAAGTCCGTGTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>HGLibB_00019</td>\n",
       "      <td>AAACGTTGTCTATTGCTGAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>HGLibB_00020</td>\n",
       "      <td>ACCCCATCTCCTATCGAGAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>HGLibB_00021</td>\n",
       "      <td>GCTTCTCGGTACTCTAGTAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGLibB_00022</td>\n",
       "      <td>GCACTCACCAATTTGTGACT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGLibB_00023</td>\n",
       "      <td>AATTTGCCCAAGTCACAAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGLibB_00024</td>\n",
       "      <td>TTTCATTTAGCACCCCAAAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>AACS</td>\n",
       "      <td>HGLibB_00025</td>\n",
       "      <td>CGCCGCCGCCATGTCCAAGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>AACS</td>\n",
       "      <td>HGLibB_00026</td>\n",
       "      <td>CTTTGTTTGCAGCAGCAATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>AACS</td>\n",
       "      <td>HGLibB_00027</td>\n",
       "      <td>CTTCCCAGGGGAAGGCAAAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>HGLibB_00028</td>\n",
       "      <td>CCTGAGAGAATCGGTATTTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>HGLibB_00029</td>\n",
       "      <td>CCAGAAATACCGATTCTCTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>HGLibB_00030</td>\n",
       "      <td>GTTATGACTTGCTGTCAAGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57998</th>\n",
       "      <td>57998</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0971</td>\n",
       "      <td>HGLibB_57999</td>\n",
       "      <td>TCATCTTACATCTGGGAGAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57999</th>\n",
       "      <td>57999</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0972</td>\n",
       "      <td>HGLibB_58000</td>\n",
       "      <td>CAAGTTTTCTGAAAGGCAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58000</th>\n",
       "      <td>58000</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0973</td>\n",
       "      <td>HGLibB_58001</td>\n",
       "      <td>AAAGATATAGCAAATTATGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58001</th>\n",
       "      <td>58001</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0974</td>\n",
       "      <td>HGLibB_58002</td>\n",
       "      <td>ACCTTACAATAAGTTATATT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58002</th>\n",
       "      <td>58002</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0975</td>\n",
       "      <td>HGLibB_58003</td>\n",
       "      <td>TGTCATGCGTCACTTAGTGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58003</th>\n",
       "      <td>58003</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0976</td>\n",
       "      <td>HGLibB_58004</td>\n",
       "      <td>AAGTGACAGATGGGCAGGCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58004</th>\n",
       "      <td>58004</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0977</td>\n",
       "      <td>HGLibB_58005</td>\n",
       "      <td>GGTCTCTGTACGGGCCGCCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58005</th>\n",
       "      <td>58005</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0978</td>\n",
       "      <td>HGLibB_58006</td>\n",
       "      <td>TTGGATATTAATTAGACATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58006</th>\n",
       "      <td>58006</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0979</td>\n",
       "      <td>HGLibB_58007</td>\n",
       "      <td>GACTTCTAGAATATAAAAGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58007</th>\n",
       "      <td>58007</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0980</td>\n",
       "      <td>HGLibB_58008</td>\n",
       "      <td>TTCCTAAGATTTTTAAGACT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58008</th>\n",
       "      <td>58008</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0981</td>\n",
       "      <td>HGLibB_58009</td>\n",
       "      <td>TAGTGGAAAATGAATATTGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58009</th>\n",
       "      <td>58009</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0982</td>\n",
       "      <td>HGLibB_58010</td>\n",
       "      <td>TTTTTAATACAAGGTAATCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58010</th>\n",
       "      <td>58010</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0983</td>\n",
       "      <td>HGLibB_58011</td>\n",
       "      <td>CGGGATGCAGCTGGAGAGGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58011</th>\n",
       "      <td>58011</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0984</td>\n",
       "      <td>HGLibB_58012</td>\n",
       "      <td>TATAACGGTAATGGGATTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58012</th>\n",
       "      <td>58012</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0985</td>\n",
       "      <td>HGLibB_58013</td>\n",
       "      <td>AAGAAGAATTGGGGATGATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58013</th>\n",
       "      <td>58013</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0986</td>\n",
       "      <td>HGLibB_58014</td>\n",
       "      <td>CAAAGGGGGAAGTGCAGGGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58014</th>\n",
       "      <td>58014</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0987</td>\n",
       "      <td>HGLibB_58015</td>\n",
       "      <td>TAAAGCAGAAGAATATACAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58015</th>\n",
       "      <td>58015</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0988</td>\n",
       "      <td>HGLibB_58016</td>\n",
       "      <td>ATTTAGTAGACATTGGGTCT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58016</th>\n",
       "      <td>58016</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0989</td>\n",
       "      <td>HGLibB_58017</td>\n",
       "      <td>GAGAAGGATGGAAATTAGAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58017</th>\n",
       "      <td>58017</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0990</td>\n",
       "      <td>HGLibB_58018</td>\n",
       "      <td>GTGTAGTCCTAGCCATGGGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58018</th>\n",
       "      <td>58018</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0991</td>\n",
       "      <td>HGLibB_58019</td>\n",
       "      <td>GCCCCGCCGCCCTCCCCTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58019</th>\n",
       "      <td>58019</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0992</td>\n",
       "      <td>HGLibB_58020</td>\n",
       "      <td>AAAGAAAGAGGAATAGTAGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58020</th>\n",
       "      <td>58020</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0993</td>\n",
       "      <td>HGLibB_58021</td>\n",
       "      <td>GAGAAGTGGGGAGCCATTGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58021</th>\n",
       "      <td>58021</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0994</td>\n",
       "      <td>HGLibB_58022</td>\n",
       "      <td>AGAAGAAAAAAATGTCTACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58022</th>\n",
       "      <td>58022</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0995</td>\n",
       "      <td>HGLibB_58023</td>\n",
       "      <td>GACTGAAATCCAAGGACTGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58023</th>\n",
       "      <td>58023</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0996</td>\n",
       "      <td>HGLibB_58024</td>\n",
       "      <td>ATATTTTATGACATAAAAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58024</th>\n",
       "      <td>58024</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0997</td>\n",
       "      <td>HGLibB_58025</td>\n",
       "      <td>TAAACAAAAAGGAAATAGTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58025</th>\n",
       "      <td>58025</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0998</td>\n",
       "      <td>HGLibB_58026</td>\n",
       "      <td>TTTCCCATGATCATTTAGTG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58026</th>\n",
       "      <td>58026</td>\n",
       "      <td>NonTargetingControlGuideForHuman_0999</td>\n",
       "      <td>HGLibB_58027</td>\n",
       "      <td>CTGAGTGAAAAATAAAAGTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58027</th>\n",
       "      <td>58027</td>\n",
       "      <td>NonTargetingControlGuideForHuman_1000</td>\n",
       "      <td>HGLibB_58028</td>\n",
       "      <td>GACAATCATGGTGAAAGCGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58028 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0                                gene_id           UID  \\\n",
       "0               0                                   A1BG  HGLibB_00001   \n",
       "1               1                                   A1BG  HGLibB_00002   \n",
       "2               2                                   A1BG  HGLibB_00003   \n",
       "3               3                                   A1CF  HGLibB_00004   \n",
       "4               4                                   A1CF  HGLibB_00005   \n",
       "5               5                                   A1CF  HGLibB_00006   \n",
       "6               6                                    A2M  HGLibB_00007   \n",
       "7               7                                    A2M  HGLibB_00008   \n",
       "8               8                                    A2M  HGLibB_00009   \n",
       "9               9                                  A2ML1  HGLibB_00010   \n",
       "10             10                                  A2ML1  HGLibB_00011   \n",
       "11             11                                  A2ML1  HGLibB_00012   \n",
       "12             12                                A3GALT2  HGLibB_00013   \n",
       "13             13                                A3GALT2  HGLibB_00014   \n",
       "14             14                                A3GALT2  HGLibB_00015   \n",
       "15             15                                 A4GALT  HGLibB_00016   \n",
       "16             16                                 A4GALT  HGLibB_00017   \n",
       "17             17                                 A4GALT  HGLibB_00018   \n",
       "18             18                                  A4GNT  HGLibB_00019   \n",
       "19             19                                  A4GNT  HGLibB_00020   \n",
       "20             20                                  A4GNT  HGLibB_00021   \n",
       "21             21                                   AAAS  HGLibB_00022   \n",
       "22             22                                   AAAS  HGLibB_00023   \n",
       "23             23                                   AAAS  HGLibB_00024   \n",
       "24             24                                   AACS  HGLibB_00025   \n",
       "25             25                                   AACS  HGLibB_00026   \n",
       "26             26                                   AACS  HGLibB_00027   \n",
       "27             27                                  AADAC  HGLibB_00028   \n",
       "28             28                                  AADAC  HGLibB_00029   \n",
       "29             29                                  AADAC  HGLibB_00030   \n",
       "...           ...                                    ...           ...   \n",
       "57998       57998  NonTargetingControlGuideForHuman_0971  HGLibB_57999   \n",
       "57999       57999  NonTargetingControlGuideForHuman_0972  HGLibB_58000   \n",
       "58000       58000  NonTargetingControlGuideForHuman_0973  HGLibB_58001   \n",
       "58001       58001  NonTargetingControlGuideForHuman_0974  HGLibB_58002   \n",
       "58002       58002  NonTargetingControlGuideForHuman_0975  HGLibB_58003   \n",
       "58003       58003  NonTargetingControlGuideForHuman_0976  HGLibB_58004   \n",
       "58004       58004  NonTargetingControlGuideForHuman_0977  HGLibB_58005   \n",
       "58005       58005  NonTargetingControlGuideForHuman_0978  HGLibB_58006   \n",
       "58006       58006  NonTargetingControlGuideForHuman_0979  HGLibB_58007   \n",
       "58007       58007  NonTargetingControlGuideForHuman_0980  HGLibB_58008   \n",
       "58008       58008  NonTargetingControlGuideForHuman_0981  HGLibB_58009   \n",
       "58009       58009  NonTargetingControlGuideForHuman_0982  HGLibB_58010   \n",
       "58010       58010  NonTargetingControlGuideForHuman_0983  HGLibB_58011   \n",
       "58011       58011  NonTargetingControlGuideForHuman_0984  HGLibB_58012   \n",
       "58012       58012  NonTargetingControlGuideForHuman_0985  HGLibB_58013   \n",
       "58013       58013  NonTargetingControlGuideForHuman_0986  HGLibB_58014   \n",
       "58014       58014  NonTargetingControlGuideForHuman_0987  HGLibB_58015   \n",
       "58015       58015  NonTargetingControlGuideForHuman_0988  HGLibB_58016   \n",
       "58016       58016  NonTargetingControlGuideForHuman_0989  HGLibB_58017   \n",
       "58017       58017  NonTargetingControlGuideForHuman_0990  HGLibB_58018   \n",
       "58018       58018  NonTargetingControlGuideForHuman_0991  HGLibB_58019   \n",
       "58019       58019  NonTargetingControlGuideForHuman_0992  HGLibB_58020   \n",
       "58020       58020  NonTargetingControlGuideForHuman_0993  HGLibB_58021   \n",
       "58021       58021  NonTargetingControlGuideForHuman_0994  HGLibB_58022   \n",
       "58022       58022  NonTargetingControlGuideForHuman_0995  HGLibB_58023   \n",
       "58023       58023  NonTargetingControlGuideForHuman_0996  HGLibB_58024   \n",
       "58024       58024  NonTargetingControlGuideForHuman_0997  HGLibB_58025   \n",
       "58025       58025  NonTargetingControlGuideForHuman_0998  HGLibB_58026   \n",
       "58026       58026  NonTargetingControlGuideForHuman_0999  HGLibB_58027   \n",
       "58027       58027  NonTargetingControlGuideForHuman_1000  HGLibB_58028   \n",
       "\n",
       "                        seq      Reserved  Count  Tag  \n",
       "0      CAATGTGACGCTGACGTGCC  HGLibB_00076      0    1  \n",
       "1      TGTCTCCGCAGGTGTCACCT           NaN      0    1  \n",
       "2      CATATGGCTGCTTCTGTCAC           NaN      0    1  \n",
       "3      GGTGCAGCATCCCAACCAGG           NaN      0    1  \n",
       "4      GGGATGCTGCACCCCCTGAA           NaN      0    1  \n",
       "5      AACTTAATAATTATGAAATT           NaN      0    1  \n",
       "6      CCCAAAGTCTTCATCCAATG           NaN      0    1  \n",
       "7      CCCACAGACGCCTCAGTCTC           NaN      0    1  \n",
       "8      TCCCAGGCAGTATATGGTTC           NaN      0    1  \n",
       "9      GACAAACCTCTCTACACCCC           NaN      0    1  \n",
       "10     GCCGGGCTGGTAATGTCACC           NaN      0    1  \n",
       "11     GTATTTCCGCATTGTCACCA           NaN      0    1  \n",
       "12     AAGTTCTGACCTGTACCCCC           NaN      0    1  \n",
       "13     GGCCGAAGGGCAGACGCCCA           NaN      0    1  \n",
       "14     AGCAAAGATAGTCAGCCCAA           NaN      0    1  \n",
       "15     CGTGCGTCGTGGGGCAGTAG           NaN      0    1  \n",
       "16     CTTCCACATGAGTGCGATCC           NaN      0    1  \n",
       "17     GAACAATGAAGTCCGTGTCC           NaN      0    1  \n",
       "18     AAACGTTGTCTATTGCTGAC           NaN      0    1  \n",
       "19     ACCCCATCTCCTATCGAGAG           NaN      0    1  \n",
       "20     GCTTCTCGGTACTCTAGTAA           NaN      0    1  \n",
       "21     GCACTCACCAATTTGTGACT           NaN      0    1  \n",
       "22     AATTTGCCCAAGTCACAAAT           NaN      0    1  \n",
       "23     TTTCATTTAGCACCCCAAAA           NaN      0    1  \n",
       "24     CGCCGCCGCCATGTCCAAGG           NaN      0    1  \n",
       "25     CTTTGTTTGCAGCAGCAATG           NaN      0    1  \n",
       "26     CTTCCCAGGGGAAGGCAAAG           NaN      0    1  \n",
       "27     CCTGAGAGAATCGGTATTTC           NaN      0    1  \n",
       "28     CCAGAAATACCGATTCTCTC           NaN      0    1  \n",
       "29     GTTATGACTTGCTGTCAAGA           NaN      0    1  \n",
       "...                     ...           ...    ...  ...  \n",
       "57998  TCATCTTACATCTGGGAGAC           NaN      0    1  \n",
       "57999  CAAGTTTTCTGAAAGGCAAT           NaN      0    1  \n",
       "58000  AAAGATATAGCAAATTATGG           NaN      0    1  \n",
       "58001  ACCTTACAATAAGTTATATT           NaN      0    1  \n",
       "58002  TGTCATGCGTCACTTAGTGC           NaN      0    1  \n",
       "58003  AAGTGACAGATGGGCAGGCG           NaN      0    1  \n",
       "58004  GGTCTCTGTACGGGCCGCCC           NaN      0    1  \n",
       "58005  TTGGATATTAATTAGACATG           NaN      0    1  \n",
       "58006  GACTTCTAGAATATAAAAGA           NaN      0    1  \n",
       "58007  TTCCTAAGATTTTTAAGACT           NaN      0    1  \n",
       "58008  TAGTGGAAAATGAATATTGA           NaN      0    1  \n",
       "58009  TTTTTAATACAAGGTAATCT           NaN      0    1  \n",
       "58010  CGGGATGCAGCTGGAGAGGA           NaN      0    1  \n",
       "58011  TATAACGGTAATGGGATTCC           NaN      0    1  \n",
       "58012  AAGAAGAATTGGGGATGATG           NaN      0    1  \n",
       "58013  CAAAGGGGGAAGTGCAGGGA           NaN      0    1  \n",
       "58014  TAAAGCAGAAGAATATACAG           NaN      0    1  \n",
       "58015  ATTTAGTAGACATTGGGTCT           NaN      0    1  \n",
       "58016  GAGAAGGATGGAAATTAGAA           NaN      0    1  \n",
       "58017  GTGTAGTCCTAGCCATGGGG           NaN      0    1  \n",
       "58018  GCCCCGCCGCCCTCCCCTCC           NaN      0    1  \n",
       "58019  AAAGAAAGAGGAATAGTAGC           NaN      0    1  \n",
       "58020  GAGAAGTGGGGAGCCATTGG           NaN      0    1  \n",
       "58021  AGAAGAAAAAAATGTCTACG           NaN      0    1  \n",
       "58022  GACTGAAATCCAAGGACTGT           NaN      0    1  \n",
       "58023  ATATTTTATGACATAAAAAT           NaN      0    1  \n",
       "58024  TAAACAAAAAGGAAATAGTT           NaN      0    1  \n",
       "58025  TTTCCCATGATCATTTAGTG           NaN      0    1  \n",
       "58026  CTGAGTGAAAAATAAAAGTT           NaN      0    1  \n",
       "58027  GACAATCATGGTGAAAGCGG           NaN      0    1  \n",
       "\n",
       "[58028 rows x 7 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table Events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def OnDoubleClick(event):\n",
    "    item = table.selection()[0]\n",
    "    value = table.item(item, 'values')\n",
    "    geneID = value[0]\n",
    "    uid = value[1]\n",
    "    sequence = value[2]\n",
    "    rc_sequence = reverseComplement(sequence)\n",
    "    \n",
    "    text_geneID.delete('1.0', tk.END)\n",
    "    text_geneID.insert('1.0', str(geneID))\n",
    "    \n",
    "    text_uid.delete('1.0', tk.END)\n",
    "    text_uid.insert('1.0', str(uid))\n",
    "    \n",
    "    text_sequence.delete('1.0', tk.END)\n",
    "    text_sequence.insert('1.0', str(sequence))\n",
    "    \n",
    "    text_rc_sequence.delete('1.0', tk.END)\n",
    "    text_rc_sequence.insert('1.0', str(rc_sequence))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def sortby(tree, col, descending):\n",
    "    \"\"\"sort tree contents when a column header is clicked on\"\"\"\n",
    "    # grab values to sort\n",
    "    data = [(tree.set(child, col), child) for child in tree.get_children('')]\n",
    "    # if the data to be sorted is numeric change to float\n",
    "    #data =  change_numeric(data)\n",
    "    # now sort the data in place\n",
    "    data.sort(reverse=descending)\n",
    "    for ix, item in enumerate(data):\n",
    "        tree.move(item[1], '', ix)\n",
    "    # switch the heading so it will sort in the opposite direction\n",
    "    tree.heading(col, command=lambda col=col: sortby(tree, col, int(not descending)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def display_in_table():\n",
    "    try:\n",
    "        for a in df.index:\n",
    "            row = df.ix[a]\n",
    "            table.insert(\"\", \"end\", \"\", values=tuple(row)) \n",
    "    except NameError:\n",
    "        messagebox.showwarning(\"No Sequences to be Displayed\", \n",
    "                               \"Sorry, there's no loaded sequences to be displayed! Please load sequence file first.\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Other Button Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def clear():\n",
    "    for i in table.get_children():\n",
    "        table.delete(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def browse():\n",
    "    start_time = time.time()\n",
    "    clear()\n",
    "    display_in_table()\n",
    "    delta_time = time.time() - start_time\n",
    "    \n",
    "    text_time.delete('1.0', tk.END)\n",
    "    text_time.insert('1.0', str(delta_time))           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonExport():   \n",
    "    if filenameSequences == '' or filenameFASTQ == '':\n",
    "        messagebox.showwarning(\"No File Loaded\", \n",
    "                               \"Sorry, no file loaded! Please choose sequence file and FASTQ file first.\")\n",
    "    else:\n",
    "        try:\n",
    "            len(df)\n",
    "            len(reads)\n",
    "            directory = filedialog.askdirectory()\n",
    "            df.to_csv(directory + '/Counts of ' + filenameSequences.split('/')[-1] + ' matched with ' + filenameFASTQ.split('/')[-1] + '.csv')\n",
    "            messagebox.showinfo(\"File Exported\", \"File of counted sequences successfully exported!\")        \n",
    "        except NameError:\n",
    "            messagebox.showwarning(\"Error: No Counted DataFrame Generated\", \n",
    "                               \"Sorry, no effective counted DataFrame generated! Please check the previous workflow.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def buttonAbout():\n",
    "    about_root=tk.Tk()\n",
    "    \n",
    "    w = 367 # width for the Tk root\n",
    "    h = 310 # height for the Tk root\n",
    "\n",
    "    # get screen width and height\n",
    "    ws = about_root.winfo_screenwidth() # width of the screen\n",
    "    hs = about_root.winfo_screenheight() # height of the screen\n",
    "\n",
    "    # calculate x and y coordinates for the Tk root window\n",
    "    x = (ws/2) - (w/2)\n",
    "    y = (hs/2) - (h/2)\n",
    "\n",
    "    # set the dimensions of the screen \n",
    "    # and where it is placed\n",
    "    about_root.geometry('%dx%d+%d+%d' % (w, h, x, y))\n",
    "    about_root.title('About Bellfort Sequence Parser')  \n",
    "    about_root.iconbitmap('dna.ico')\n",
    "\n",
    "    label_author=tk.Label(about_root,text='Bellfort Sequence Parser Version 1.0', font=('tahoma', 9))\n",
    "    label_author.place(x=90,y=30)\n",
    "\n",
    "    label_author=tk.Label(about_root,text='Copyright (C) 2016', font=('tahoma', 9))\n",
    "    label_author.place(x=125,y=60)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Chen Lab', font=('tahoma', 9))\n",
    "    label_author.place(x=150,y=90)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Human Genome Sequencing Center', font=('tahoma', 9))\n",
    "    label_author.place(x=80,y=120)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Department of Molecular and Human Genetics', font=('tahoma', 9))\n",
    "    label_author.place(x=50,y=150)\n",
    "    \n",
    "    label_author=tk.Label(about_root,text='Baylor College of Medicine', font=('tahoma', 9))\n",
    "    label_author.place(x=110,y=180)\n",
    "   \n",
    "\n",
    "    button_okay=ttk.Button(about_root, width=15, text='OK', command=about_root.destroy)\n",
    "    button_okay.place(x=130, y=235)\n",
    "\n",
    "    about_root.mainloop()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Flow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "headers = ['gene_id', 'UID', 'seq', 'Reserved', 'count', 'tag']\n",
    "header_widths = [280, 150, 350, 100, 80, 100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "root = tk.Tk()\n",
    "\n",
    "indicator_preprocess = 0\n",
    "indicator_loadSequences = 0\n",
    "indicator_matchAll = 0\n",
    "filenameSequences = ''\n",
    "filenameFASTQ = ''\n",
    "recordNum = 0\n",
    "count = 0\n",
    "\n",
    "root.geometry(\"{0}x{1}+0+0\".format(root.winfo_screenwidth(), root.winfo_screenheight()))\n",
    "#root.attributes('-fullscreen', True)\n",
    "root.title('Bellfort Sequence Parser')\n",
    "root.iconbitmap('dna.ico')\n",
    "\n",
    "\n",
    "# Multicolumn Listbox/////////////////////////////////////////////////////////////////////////////\n",
    "table = ttk.Treeview(height=\"20\", columns=headers, selectmode=\"extended\")\n",
    "table.pack(padx=10, pady=20, ipadx=1200, ipady=130)\n",
    "\n",
    "i = 1\n",
    "for header in headers:\n",
    "    table.heading('#'+str(i), text=header.title(), anchor=tk.W, command=lambda c=header: sortby(table, c, 0))\n",
    "    table.column('#'+str(i), stretch=tk.NO, minwidth=0, width=tkf.Font().measure(header.title())+header_widths[i-1]) \n",
    "    i+=1    \n",
    "table.column('#0', stretch=tk.NO, minwidth=0, width=0)\n",
    "\n",
    "table.bind(\"<Double-1>\", OnDoubleClick)\n",
    "#///////////////////////////////////////////////////////////////////////////////////////////\n",
    "\n",
    "# Scrollbar////////////////////////////////////////////////////////////////////////////////////////\n",
    "vsb = ttk.Scrollbar(table, orient=\"vertical\",  command = table.yview)\n",
    "hsb = ttk.Scrollbar(table, orient=\"horizontal\", command = table.xview)\n",
    "## Link scrollbars activation to top-level object\n",
    "table.configure(yscrollcommand=vsb.set, xscrollcommand=hsb.set)\n",
    "## Link scrollbar also to every columns\n",
    "map(lambda col: col.configure(yscrollcommand=vsb.set,xscrollcommand=hsb.set), table)\n",
    "vsb.pack(side = tk.RIGHT, fill = tk.Y)\n",
    "hsb.pack(side = tk.BOTTOM, fill = tk.X)        \n",
    "\n",
    "#//////////////////////////////////////////////////////////////////////////////////////////////\n",
    "y0 =370\n",
    "y1 = 410\n",
    "y2 = 480\n",
    "y3 = 520\n",
    "y4 = 580\n",
    "y5 = 615\n",
    "y6 = 655\n",
    "y7 = 695\n",
    "# Text /////////////////////////////////////////////////////////////////////////////////////\n",
    "text_recordNum=tk.Text(root, width=18, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_recordNum.place(x=830, y=y0)\n",
    "label_recordNum=tk.Label(root, text='records', font=('tahoma', 9))\n",
    "label_recordNum.place(x=990,y=y0)\n",
    "\n",
    "text_fileSequences=tk.Text(root, width=50, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileSequences.place(x=60, y=y0)\n",
    "\n",
    "text_fileFASTQ=tk.Text(root, width=36, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_fileFASTQ.place(x=60, y=y4)\n",
    "\n",
    "text_count=tk.Text(root, width=16, height=1, font=('tahoma', 9), bd=2)\n",
    "text_count.place(x=1000, y=y3)\n",
    "label_count=tk.Label(root, text='Count:', font=('tahoma', 9))\n",
    "label_count.place(x=940,y=y3)\n",
    "\n",
    "text_geneID=tk.Text(root, width=20, height=1, font=('tahoma', 9), bd=2)\n",
    "text_geneID.place(x=140, y=y2)\n",
    "label_geneID=tk.Label(root, text='Gene ID:', font=('tahoma', 9))\n",
    "label_geneID.place(x=60,y=y2)\n",
    "\n",
    "text_uid=tk.Text(root, width=20, height=1, font=('tahoma', 9), bd=2)\n",
    "text_uid.place(x=390, y=y2)\n",
    "label_uid=tk.Label(root, text='UID:', font=('tahoma', 9))\n",
    "label_uid.place(x=340,y=y2)\n",
    "\n",
    "text_sequence=tk.Text(root, width=38, height=1, font=('tahoma', 9), bd=2)\n",
    "text_sequence.place(x=680, y=y2)\n",
    "label_sequence=tk.Label(root, text='Sequence:', font=('tahoma', 9))\n",
    "label_sequence.place(x=600,y=y2)\n",
    "\n",
    "text_rc_sequence=tk.Text(root, width=38, height=1, font=('tahoma', 9), bd=2)\n",
    "text_rc_sequence.place(x=1000, y=y2)\n",
    "\n",
    "text_sequence_len=tk.Text(root, width=5, height=1, font=('tahoma', 9), bd=2)\n",
    "text_sequence_len.place(x=1010, y=y5)\n",
    "label_sequence_len=tk.Label(root, text='nts', font=('tahoma', 9))\n",
    "label_sequence_len.place(x=1065,y=y5)\n",
    "text_sequence_len.delete('1.0', tk.END)\n",
    "text_sequence_len.insert('1.0', str(20))\n",
    "\n",
    "text_readNumDivided=tk.Text(root, width=13, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_readNumDivided.place(x=295, y=y3+10)\n",
    "label_readNumDivided1=tk.Label(root, text='by', font=('tahoma', 9))\n",
    "label_readNumDivided1.place(x=255,y=y3+10)\n",
    "label_readNumDivided2=tk.Label(root, text='reads', font=('tahoma', 9))\n",
    "label_readNumDivided2.place(x=420,y=y3+10)\n",
    "text_readNumDivided.delete('1.0', tk.END)\n",
    "text_readNumDivided.insert('1.0', str(2000000))\n",
    "\n",
    "text_readNum=tk.Text(root, width=22, height=1, font=('tahoma', 9), bd=2, wrap='none')\n",
    "text_readNum.place(x=400, y=y6)\n",
    "label_readNum=tk.Label(root, text='reads', font=('tahoma', 9))\n",
    "label_readNum.place(x=590,y=y6)\n",
    "\n",
    "text_time=tk.Text(root, width=15, height=1, font=('tahoma', 9), bd=2)\n",
    "text_time.place(x=115, y=y7)\n",
    "label_time=tk.Label(root, text='Time:', font=('tahoma', 9))\n",
    "label_time.place(x=60,y=y7)\n",
    "label_seconds=tk.Label(root, text='second(s)', font=('tahoma', 9))\n",
    "label_seconds.place(x=250,y=y7)\n",
    "\n",
    "# ProgressBar /////////////////////////////////////////////////////////////////////////////\n",
    "progressbar_loadSequences = ttk.Progressbar(root, length=200, maximum=100, mode='determinate')\n",
    "progressbar_loadSequences.place(x=500,y=y0)\n",
    "\n",
    "progressbar_loadFASTQ = ttk.Progressbar(root, length=300, mode='indeterminate')\n",
    "progressbar_loadFASTQ.place(x=400,y=y4)\n",
    "\n",
    "progressbar = ttk.Progressbar(root, length=550, maximum=1000000, mode='determinate')\n",
    "progressbar.place(x=760,y=y4)\n",
    "\n",
    "# Button /////////////////////////////////////////////////////////////////////////////////\n",
    "button_browseSequences = ttk.Button(root, text=\"Browse sgRNA...\", width=20, command=buttonBrowseSequences)\n",
    "button_browseSequences.place(x=60, y=y1)\n",
    "\n",
    "button_loadSequences = ttk.Button(root, text=\"Load sgRNA\", width=20, command=loadSequences)\n",
    "button_loadSequences.place(x=500, y=y1)\n",
    "\n",
    "button_loadHalfMatchedSequences = ttk.Button(root, text=\"Load Half Matched sgRNA\", width=30, command=buttonLoadHalfMatchedSequences)\n",
    "button_loadHalfMatchedSequences.place(x=800, y=y1)\n",
    "\n",
    "button_clear = ttk.Button(root, text=\"Clear\", width=20, command=clear)\n",
    "button_clear.place(x=1180, y=y1)\n",
    "\n",
    "button_refresh = ttk.Button(root, text=\"Browse\", width=20, command=browse)\n",
    "button_refresh.place(x=1180, y=y0)\n",
    "\n",
    "button_browseFASTQ = ttk.Button(root, text=\"Browse FASTQ...\", width=20, command=buttonBrowseFASTQ)\n",
    "button_browseFASTQ.place(x=60, y=y5)\n",
    "\n",
    "button_divideFASTQ = ttk.Button(root, text=\"Divide FASTQ\", width=20, command=lambda:start_divideFASTQ_thread(None))\n",
    "button_divideFASTQ.place(x=60, y=y3+10)\n",
    "\n",
    "button_loadFASTQ = ttk.Button(root, text=\"Load FASTQ\", width=20, command=lambda:start_loadFASTQ_thread(None))\n",
    "button_loadFASTQ.place(x=400, y=y5)\n",
    "\n",
    "button_preprocessFASTQ = ttk.Button(root, text=\"Preprocess FASTQ\", width=20, command=lambda:start_preprocess_thread(None))\n",
    "button_preprocessFASTQ.place(x=760, y=y5)\n",
    "\n",
    "button_match = ttk.Button(root, text=\"Match\", width=20, command=buttonMatch)\n",
    "button_match.place(x=680, y=y3)\n",
    "\n",
    "button_matchAll = ttk.Button(root, text=\"Match All\", width=20, command=lambda:start_matchAll_thread(None))\n",
    "button_matchAll.place(x=1180, y=y5)\n",
    "\n",
    "button_about = ttk.Button(root, text=\"About\", width=20, command=buttonAbout)\n",
    "button_about.place(x=980, y=y7)\n",
    "\n",
    "button_export = ttk.Button(root, text=\"Export\", width=20, command=buttonExport)\n",
    "button_export.place(x=720, y=y7)\n",
    "\n",
    "button_exit = ttk.Button(root, text=\"Exit\", width=20, command=root.destroy)\n",
    "button_exit.place(x=1180, y=y7)\n",
    "\n",
    "root.bind('<Return>', start_preprocess_thread)\n",
    "root.bind('<Return>', start_loadFASTQ_thread)\n",
    "root.bind('<Return>', start_divideFASTQ_thread)\n",
    "root.bind('<Return>', start_matchAll_thread)\n",
    "\n",
    "root.mainloop()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
